cni/network/network.go

```diff
@@ -453,274 +527,221 @@ func (plugin *NetPlugin) Add(args *cniSkel.CmdArgs) error {
+	}
+
+	policies := cni.GetPoliciesFromNwCfg(nwCfg.AdditionalArgs)
+	// moved to addIpamInvoker
+	// sendEvent(plugin, fmt.Sprintf("Allocated IPAddress from ipam interface: %+v", ipamAddResult.PrettyString()))
+
+	defer func() { //nolint:gocritic
+		if err != nil {
+			// for multi-tenancies scenario, CNI is not supposed to invoke CNS for cleaning Ips
+			// if nwCfg.Multitenancy is false or ipam type is not azure cns
+			if !(nwCfg.MultiTenancy && nwCfg.IPAM.Type == network.AzureCNS) {
+				for _, ifInfo := range ipamAddResult.interfaceInfo {
+					// This used to only be called for infraNIC, test if this breaks scenarios
+					// If it does then will have to search for infraNIC
+					if ifInfo.NICType == cns.InfraNIC {
+						plugin.cleanupAllocationOnError(ifInfo.IPConfigs, nwCfg, args, options)
+					}
 				}
 			}
-		}()
-
-		// Create network
-		if nwInfoErr != nil {
-			// Network does not exist.
-			logger.Info("Creating network", zap.String("networkID", networkID))
-			sendEvent(plugin, fmt.Sprintf("[cni-net] Creating network %v.", networkID))
-			// opts map needs to get passed in here
-			if nwInfo, err = plugin.createNetworkInternal(networkID, policies, ipamAddConfig, ipamAddResult); err != nil {
-				logger.Error("Create network failed", zap.Error(err))
-				return err
-			}
-			logger.Info("Created network",
-				zap.String("networkId", networkID),
-				zap.String("subnet", ipamAddResult.hostSubnetPrefix.String()))
-			sendEvent(plugin, fmt.Sprintf("[cni-net] Created network %v with subnet %v.", networkID, ipamAddResult.hostSubnetPrefix.String()))
 		}
+	}()
+
+	epInfos := []*network.EndpointInfo{}
+	infraSeen := false
+	i := 0
+	for key := range ipamAddResult.interfaceInfo {
+		ifInfo := ipamAddResult.interfaceInfo[key]
 
 		natInfo := getNATInfo(nwCfg, options[network.SNATIPKey], enableSnatForDNS)
+		networkID, _ := plugin.getNetworkID(args.Netns, &ifInfo, nwCfg)
 
-		createEndpointInternalOpt := createEndpointInternalOpt{
+		createEpInfoOpt := createEpInfoOpt{
 			nwCfg:            nwCfg,
-			cnsNetworkConfig: ipamAddResult.ncResponse,
+			cnsNetworkConfig: ifInfo.NCResponse,
 			ipamAddResult:    ipamAddResult,
 			azIpamResult:     azIpamResult,
 			args:             args,
-			nwInfo:           &nwInfo,
 			policies:         policies,
-			endpointID:       endpointID,
 			k8sPodName:       k8sPodName,
 			k8sNamespace:     k8sNamespace,
 			enableInfraVnet:  enableInfraVnet,
 			enableSnatForDNS: enableSnatForDNS,
 			natInfo:          natInfo,
-		}
+			networkID:        networkID,
+
+			ifInfo:        &ifInfo,
+			ipamAddConfig: &ipamAddConfig,
+			ipv6Enabled:   ipamAddResult.ipv6Enabled,
 
-		var epInfo network.EndpointInfo
-		epInfo, err = plugin.createEndpointInternal(&createEndpointInternalOpt)
+			infraSeen: &infraSeen,
+			idx:       i,
+		}
+		var epInfo *network.EndpointInfo
```

Comment:
NIT: let's move this to line 376 as the global variable

Reply:
why? I don't want this variable to be visible/used prior to this line by accident